### PR TITLE
Add feature to add user selected log path

### PR DESCRIPTION
**Pull request by vivek-kumar-004**
# Changes

Add API to set the log file from selected options from UI
Bind the values in the ui component to call the api when checkbox selection is changed
In the getSettings api, add another field to return the logFile

## TODOs:
- [ ] Values retained on page refresh
- [ ] Fetch the radio button options dynamically from list
- [ ] Add another component for radio button

### CODE REVIEW

File: lib/Log/LogIteratorFactory.php [hunk 1]
```diff
@@ -23,43 +23,76 @@
...
+
 		throw new \Exception('Can\'t find log class');
 	}
+
+	private function getIterator(
+		string $logFilePath,
+		string $dateFormat,
+		string $timezone,
+		array $levels
+	) {
```
Comment (Abijeet) [inline lib/Log/LogIteratorFactory.php:87]: return type

File: lib/Log/LogIteratorFactory.php [hunk 2]
```diff
@@ -23,43 +23,76 @@
...
+	{
 		$levels = str_split($levelsString);
 		$levels = array_map(function ($level) {
 			return $level === '1';
 		}, $levels);
 		$dateFormat = $this->config->getSystemValue('logdateformat', \DateTime::ATOM);
 		$timezone = $this->config->getSystemValue('logtimezone', 'UTC');
+
+		$selectedLogFile = $this->config->getAppValue('logreader', 'logFile');
+		if ($selectedLogFile && $selectedLogFilePath = $this->systemConfig->getValue('logAdditional', $selectedLogFile)[$selectedLogFile]) {
```
Comment (Abijeet) [inline lib/Log/LogIteratorFactory.php:60]: is this not available from config itself? Why do we need systemconfig?

File: lib/Log/LogIteratorFactory.php [hunk 3]
```diff
@@ -23,43 +23,76 @@
...
+	public function getLogIterator($levelsString)
+	{
 		$levels = str_split($levelsString);
 		$levels = array_map(function ($level) {
 			return $level === '1';
 		}, $levels);
 		$dateFormat = $this->config->getSystemValue('logdateformat', \DateTime::ATOM);
 		$timezone = $this->config->getSystemValue('logtimezone', 'UTC');
+
+		$selectedLogFile = $this->config->getAppValue('logreader', 'logFile');
```
Comment (Abijeet) [inline lib/Log/LogIteratorFactory.php:59]:
```
$this->systemConfig->getValue('logAdditional', $selectedLogFile)[$selectedLogFile]
```
Check if it exists

File: lib/Log/LogIteratorFactory.php [hunk 4]
```diff
@@ -23,43 +23,76 @@
...
+		array $levels
+	) {
+		$handle = fopen($logFilePath, 'rb');
+		if ($handle) {
+			$iterator = new LogIterator($handle, $dateFormat, $timezone);
+			return new \CallbackFilterIterator($iterator, function ($logItem) use ($levels) {
+				return $logItem && isset($levels[$logItem['level']]) && $levels[$logItem['level']];
+			});
+		} else {
+			throw new \Exception("Error while opening " . $logFilePath);
```
Comment (Abijeet) [inline lib/Log/LogIteratorFactory.php:95]: Why are you using double quotes?

File: lib/Controller/LogController.php [hunk 5]
```diff
@@ -167,7 +168,13 @@ public function setLive($live) {
 		$this->config->setAppValue('logreader', 'live', $live);
 	}
 
-	public function setLevels($levels) {
+	public function setLogFile($logFile)
```
Comment (Abijeet) [inline lib/Controller/LogController.php:171]: Lets specify return type in our code.

**Comment by vivek-kumar-004** [comment]
Closing this PR as this PR is created on master branch

File: lib/Log/LogIteratorFactory.php (hunk 1 above)
Comment (vivek-kumar-004) [reply lib/Log/LogIteratorFactory.php]: Added return type

File: lib/Log/LogIteratorFactory.php (hunk 2 above)
Comment (vivek-kumar-004) [reply lib/Log/LogIteratorFactory.php]: config is used to set the type of log in the db, 
the actual log file path is stored in `config.php`

File: lib/Log/LogIteratorFactory.php (hunk 3 above)
Comment (vivek-kumar-004) [reply lib/Log/LogIteratorFactory.php:59]: Added the check

File: lib/Log/LogIteratorFactory.php (hunk 4 above)
Comment (vivek-kumar-004) [reply lib/Log/LogIteratorFactory.php]: Changed to single quotes

File: lib/Controller/LogController.php (hunk 5 above)
Comment (vivek-kumar-004) [reply lib/Controller/LogController.php]: Added return type

**Comment by vivek-kumar-004** [comment]
Closing this PR as it is created on base nextcloud master

**Pull request closed without merging**